test/provider_test.c

```diff
@@ -19,41 +19,69 @@ static OSSL_PARAM greeting_request[] = {
     { NULL, 0, NULL, 0, 0 }
 };
 
-static int test_provider(const char *name)
+static int test_provider(OSSL_LIB_CTX **libctx, const char *name)
 {
     OSSL_PROVIDER *prov = NULL;
     const char *greeting = NULL;
     char expected_greeting[256];
+    int ok;
 
     BIO_snprintf(expected_greeting, sizeof(expected_greeting),
                  "Hello OpenSSL %.20s, greetings from %s!",
                  OPENSSL_VERSION_STR, name);
 
-    return
-        TEST_ptr(prov = OSSL_PROVIDER_load(NULL, name))
+    ok =
+        TEST_ptr(prov = OSSL_PROVIDER_load(*libctx, name))
         && TEST_true(OSSL_PROVIDER_get_params(prov, greeting_request))
         && TEST_ptr(greeting = greeting_request[0].data)
         && TEST_size_t_gt(greeting_request[0].data_size, 0)
         && TEST_str_eq(greeting, expected_greeting)
         && TEST_true(OSSL_PROVIDER_unload(prov));
+
+    /*
+     * We must free the libctx to force the provider to really be unloaded from
+     * memory
+     */
+    OSSL_LIB_CTX_free(*libctx);
+    *libctx = NULL;
+
+    /*
+     * Check that errors created by the now unloaded provider can still be
```

Comment:
Maybe you should check that there is actually some error or even the actual error we expect to be there on the stack? Otherwise the test might be bypassed if there was some ERR_clear_errors() call somewhere.

Reply:
Done